### PR TITLE
Adds tests for demo configuration java tool

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
This PR adds tests associated with the demo-configuration tool.

## Issues Resolved:
 - Resolves https://github.com/opensearch-project/security/issues/3636

### Testing
- Unit Testing
- Integration Testing

### Check List
- [x] New functionality includes testing
~- [ ] New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by DarshitChanpura** [comment]
continuing the discussion here from #3777 :

> You've got an methods showHelp() and readOptions() both of these do not share any common code, there are many libraries such as those used by the [SecurityAdmin](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/tools/SecurityAdmin.java#L200) that handle these use cases. Then rather than checking a string output you could verify the collection - or frankly not even verify that since its just configuration at that point.

Tests should verify something that can break the API / user contract. Consider the usage of the code and work backward for what should be tested.

File: src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java [hunk 1]
```diff
@@ -45,6 +45,8 @@ public class SecurityRestTestCase extends OpenSearchRestTestCase {
 
     private static final String CERT_FILE_DIRECTORY = "sanity-tests/";
 
+    static boolean useAdminAsPassword = false;
```
Comment (peternied) [inline src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java:48]: Adding static override is a code smell, instead could the test set the property of `password` to be admin?

File: src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    @BeforeClass
+    public static void setUpInvalidAdminCredentials() {
+        useAdminAsPassword = true;
+    }
+
+    @AfterClass
+    public static void restoreValidAdminCredentials() {
+        useAdminAsPassword = false;
+    }
```
Comment (peternied) [inline src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java:36]: This is hard to follow, instead could we encode the username/password as a basic auth header in the way that gets a 401 and then the other where it works as expected?

File: src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java (hunk 1 above)
Comment (peternied) [reply src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java:48]: We should delete this

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java [hunk 3]
```diff
@@ -34,6 +38,13 @@ public void testSecurityPluginInstallation() throws Exception {
         verifyPluginInstallationOnAllNodes();
     }
 
+    @Test
+    public void testAdminCredentials_customStrongPassword_shouldSucceed() throws Exception {
+        Response response = client().performRequest(new Request("GET", ""));
```
Comment (peternied) [inline src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:43]: This test doesn't set or verify the strength of the password, can we change the name or make the test assert something about the state of the system?

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java [hunk 4]
```diff
@@ -34,6 +38,13 @@ public void testSecurityPluginInstallation() throws Exception {
...
+        MatcherAssert.assertThat(response.getStatusLine().getStatusCode(), is(equalTo(200)));
+        MatcherAssert.assertThat(response.getStatusLine().getReasonPhrase(), is(equalTo("OK")));
```
Comment (peternied) [inline src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:45]: nit: use static import of `assertThat(...)`

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                Instant expiry = expiryDate.toInstant();
+
+                assertThat(isExpiryAtLeastAYearLater(expiry), is(equalTo(true)));
+                assertThat(x509Certificate.getSigAlgName(), is(equalTo("SHA256withRSA")));
+            } else {
+                fail("Certificate is invalid. Expected X.509 certificate.");
+            }
+        } catch (Exception e) {
+            fail("Error checking certificate validity: " + e.getMessage());
+        }
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java:116]: nit: no need to catch, `throws Exception` in the method declaration

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            Certificate certificate = cf.generateCertificate(certInputStream);
+
+            if (certificate instanceof X509Certificate) {
+                X509Certificate x509Certificate = (X509Certificate) certificate;
+                x509Certificate.checkValidity();
+
+                Date expiryDate = x509Certificate.getNotAfter();
+                Instant expiry = expiryDate.toInstant();
+
+                assertThat(isExpiryAtLeastAYearLater(expiry), is(equalTo(true)));
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java:109]: nit: you can use `org.hamcrest.Matchers.greaterThan` to verify the time, then you'll get a meaningful error if the check fails.  Checking `equalTo(true)` doesn't produce a useful output

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        try {
+            String pemContent = readPEMFile(keyPath);
+
+            String base64Data = pemContent.replaceAll("-----BEGIN PRIVATE KEY-----|-----END PRIVATE KEY-----", "").replaceAll("\\s", "");
+
+            byte[] keyBytes = Base64.getDecoder().decode(base64Data);
+            KeyFactory kf = KeyFactory.getInstance("RSA");
+            PrivateKey key = kf.generatePrivate(new PKCS8EncodedKeySpec(keyBytes));
+            assertThat(key.getFormat(), is(equalTo("PKCS#8")));
+            assertThat(key.getAlgorithm(), is(equalTo("RSA")));
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java:138]: 🔥

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+        for (Certificates cert : certificatesArray) {
+            String certFilePath = installer.OPENSEARCH_CONF_DIR + File.separator + cert.getFileName();
+            File certFile = new File(certFilePath);
+            assertThat(certFile.exists(), is(equalTo(true)));
+            assertThat(certFile.canRead(), is(equalTo(true)));
+
+            if (certFilePath.endsWith("-key.pem")) {
+                checkPrivateKeyValidity(certFilePath);
+                return;
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java:77]: This can exit early before everything is verified.  IMO there should be an expected number of cert generated and check should be asserted against (no loop)

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java (hunk 4 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:45]: for some unknown reason, assertThat without inline import (with static import) always refers to JUnit assertThat instead of MatcherAssert's assertThat. That is why I had to resort to inline imports

File: src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java (hunk 1 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java]: refactored.

File: src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java]: since the client is build before the class is invoked, the only way to build client with invalid credentials is using @BeforeClass and then setting invalid admin password.

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java (hunk 3 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java]: the purpose of this test is to verify that admin is called with correct credentials and those correct credentials are assumed to have strong password which was used during cluster formation.

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java (hunk 5 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java]: done

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java (hunk 6 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java]: agreed and done.

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java (hunk 8 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java]: refactored to verify correct behavior. I think I meant to write break instead of return here. Also added a check for expected number of certificate files.
I've kept it a loop so that any changes to Certificates.java would not trigger re-write this completely. Rather the test would fail if number of certificates files was now greater or less than 5.

**Comment by DarshitChanpura** [comment]
ty @peternied for the feedback. I've addressed all comments and have refactored some tests.

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java (hunk 4 above)
Comment (peternied) [reply src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:45]: Aside: Can you dig into this outside of this PR? Having unknown behavior in codebase seems like a "Bad Thing" (tm)

File: src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java (hunk 2 above)
Comment (peternied) [reply src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java]: This seems to be due to the construction of the rest client, could you create a follow up issue? Not being able to change credentials during a test seems like it will hamper many commonly tested scenarios in our codebase.

**Review by peternied — APPROVED** [review]

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java (hunk 4 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:45]: sounds good. Created an issue: https://github.com/opensearch-project/security/issues/3832

**Review by scrawfor99 — APPROVED** [review]
Looks good to me! Thanks DC!
